runtime/common/addresslocation.go

```diff
@@ -0,0 +1,158 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package common
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+const AddressLocationPrefix = "A"
+
+// AddressLocation is the location of a contract/contract interface at an address
+//
+type AddressLocation struct {
+	Address Address
+	Name    string
+}
+
+func (l AddressLocation) String() string {
+	if l.Name == "" {
+		return l.Address.String()
+	}
+
+	return fmt.Sprintf(
+		"%s.%s",
+		l.Address.String(),
+		l.Name,
+	)
+}
+
+func (l AddressLocation) ID() LocationID {
+	if l.Name == "" {
+		return NewLocationID(
+			AddressLocationPrefix,
+			l.Address.Hex(),
+		)
+	}
+
+	return NewLocationID(
+		AddressLocationPrefix,
+		l.Address.Hex(),
+		l.Name,
+	)
+}
+
+func (l AddressLocation) TypeID(qualifiedIdentifier string) TypeID {
+	return NewTypeID(
+		AddressLocationPrefix,
+		l.Address.Hex(),
+		qualifiedIdentifier,
+	)
+}
+
+func (l AddressLocation) QualifiedIdentifier(typeID TypeID) string {
+	pieces := strings.SplitN(string(typeID), ".", 3)
+
+	if len(pieces) < 3 {
+		return ""
+	}
+
+	return pieces[2]
+}
+
+func (l AddressLocation) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&struct {
+		Type    string
+		Address string
+		Name    string
+	}{
+		Type:    "AddressLocation",
+		Address: l.Address.ShortHexWithPrefix(),
+		Name:    l.Name,
+	})
+}
+
+func init() {
+	RegisterTypeIDDecoder(
+		AddressLocationPrefix,
+		func(typeID string) (location Location, qualifiedIdentifier string, err error) {
+			return decodeAddressLocationTypeID(typeID)
+		},
+	)
+}
+
+func decodeAddressLocationTypeID(typeID string) (AddressLocation, string, error) {
+
+	const errorMessagePrefix = "invalid address location type ID"
+
+	newError := func(message string) (AddressLocation, string, error) {
+		return AddressLocation{}, "", fmt.Errorf("%s: %s", errorMessagePrefix, message)
+	}
+
+	if typeID == "" {
+		return newError("missing prefix")
+	}
+
+	parts := strings.SplitN(typeID, ".", 4)
+
+	pieceCount := len(parts)
+	switch pieceCount {
+	case 1:
+		return newError("missing location")
+	case 2:
+		return newError("missing qualified identifier")
+	}
+
+	prefix := parts[0]
+
+	if prefix != AddressLocationPrefix {
+		return AddressLocation{}, "", fmt.Errorf(
+			"%s: invalid prefix: expected %q, got %q",
+			errorMessagePrefix,
+			AddressLocationPrefix,
+			prefix,
+		)
+	}
+
+	address, err := hex.DecodeString(parts[1])
+	if err != nil {
+		return AddressLocation{}, "", fmt.Errorf(
+			"%s: invalid address: %w",
+			errorMessagePrefix,
+			err,
+		)
+	}
+
+	name := parts[2]
+	qualifiedIdentifier := name
+
+	if pieceCount > 3 {
+		qualifiedIdentifier = fmt.Sprintf("%s.%s", qualifiedIdentifier, parts[3])
+	}
```

Comment:
A comment explaining these two cases would be handy. I'm guessing one case is `A.<address>.<variableName>` and the other is `A.<address>.<contractName>.<variableName>`? 

Or maybe we can name the different parts in each case like:
```suggestion
	var qualifiedIdentifier
	switch pieceCount {
	case 3:
		varName := parts[2]
		qualifiedIdentifier = varName
	case 4:
		contractName := parts[2]
		varName := parts[3]
		qualifiedIdentifier = fmt.Sprintf("%s.%s", contractName, varName)
	default:
	// return error?
	}

```

Reply:
Good point, I tried to explain the decoding by adding comments and making cases more explicit in 1cf82ef.